### PR TITLE
<link rel> Attribute additions, link relationship additions, and <noscript> element.

### PR DESCRIPTION
**Pull request by rpinz**
Some additions that were necessary to implement a complete web page render test of [HTML5 Boilerplate](https://github.com/h5bp/html5-boilerplate).  I believe this test is a valuable addition to both show others in code a complete (yet simple) page implementation and serve as a full functionality test.  It's also a great copy/paste starting point, and potentially a template for publish.

### CODE REVIEW

File: Sources/Plot/API/HTMLElements.swift [hunk 1]
```diff
@@ -247,6 +247,12 @@ public extension Node where Context: HTML.BodyContext {
         .element(named: "i", nodes: nodes)
     }
 
+    /// Add an `<!--[if IE]>...<![endif]-->` HTML element within the current context.
+    /// - parameter nodes: The element's attributes and child elements.
+    static func ifIE(_ nodes: Node<Context>...) -> Node {
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLElements.swift]: Not sure if I think this is worth including in Plot itself, as it can be achieved using an HTML comment. Also, do people still use this in practice? I thought we were past the days of Internet Explorer-special casing?

File: Sources/Plot/API/HTMLAttributes.swift [hunk 2]
```diff
@@ -128,6 +128,30 @@ public extension Node where Context: HTMLLinkableContext {
     }
 }
 
+public extension Attribute where Context: HTMLLinkableContext {
+    /// Assign a hreflang string to this element.
+    /// - parameter hreflang: The hreflang of the link to assign.
+    static func hreflang(_ hreflang: String) -> Attribute {
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: Shouldn't this accept a `Language` value, rather than a string?

File: Sources/Plot/API/HTMLAttributes.swift [hunk 3]
```diff
@@ -128,6 +128,30 @@ public extension Node where Context: HTMLLinkableContext {
...
+        Attribute(name: "hreflang", value: hreflang)
+    }
+}
+
+public extension Attribute where Context: HTMLLinkableContext {
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: I think this should be constrained by `Context == HTML.LinkContext`, since it can only be used on `<link>` elements.

File: Sources/Plot/API/HTMLAttributes.swift [hunk 4]
```diff
@@ -128,6 +128,30 @@ public extension Node where Context: HTMLLinkableContext {
...
+    /// - parameter sizes: The icon sizes of the link to assign.
+    static func sizes(_ sizes: String) -> Attribute {
+        Attribute(name: "sizes", value: sizes)
+    }
+}
+
+public extension Attribute where Context: HTMLLinkableContext {
+    /// Assign an icon color string to this element.
+    /// - parameter color: The icon color of the link to assign.
+    static func color(_ color: String) -> Attribute {
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: Same comment here as above, and these two extensions can be merged into one.

File: Tests/PlotTests/Assertions.swift [hunk 5]
```diff
@@ -10,11 +10,12 @@ import Plot
 func assertEqualHTMLContent(
     _ document: HTML,
     _ content: String,
+    _ expected: String? = nil,
     file: StaticString = #file,
     line: UInt = #line
 ) {
     let html = document.render()
-    let expectedPrefix = "<!DOCTYPE html><html>"
+    let expectedPrefix = expected ?? "<!DOCTYPE html><html>"
```
Comment (JohnSundell) [inline Tests/PlotTests/Assertions.swift]: What's the purpose of this change?

File: Sources/Plot/API/HTMLElements.swift (hunk 1 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLElements.swift]: After some thought, I do have to agree, but for different rationale.  I added it because my attempts to add [Conditional Comments](https://en.wikipedia.org/wiki/Conditional_comment) inside .comment() failed to be parsed properly by at least Firefox.  The effort required to fully implement conditional comments within the DSL would be more than this commit, and likely not worth the effort for this limited edge case since it can still be implemented with .raw() so I removed the commits.

As to your questions, it is still used in practice, but the use cases are more limited for sure.  Many users still do use EOL old versions of IE for some reason.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 3 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: Updated.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 4 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: Updated.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 2 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: Done.  And added "x-default" to language enum.

File: Tests/PlotTests/Assertions.swift (hunk 5 above)
Comment (rpinz) [reply Tests/PlotTests/Assertions.swift]: That was a modification for the HTML5 Boilerplate test, `.html()` has `.class()` and `.lang()` attributes which renders to "`<!DOCTYPE html><html class="no-js" lang="en">`" so a closed "`<html>`" would cause failure.  This was just a minimal method (and effort) to allow for customization.  Perhaps the "`<html>`" should be removed from the `expectedPrefix` test?  Some variation could be useful still because it is additive, defaults to nil and allows for testing attributes on the "`<html>`" element.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 2 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: More long-term and for complete support here for sites or pages that need to have [localized language versions](https://support.google.com/webmasters/answer/189077?hl=en#language-codes) of a page indexed, country support ([ISO 3166-1](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2)) should be added to hreflang as well. Which would allow for "en-GB" for instance.  I just pushed an update with a potential solution.  The biggest issue I see is the coutry names, many are multiple words, so I chose to just basically map lower-case country inputs to uppercase: `hreflang(.english, .us)` which renders `hreflang="en-US"`.

File: Tests/PlotTests/Assertions.swift (hunk 5 above)
Comment (rpinz) [reply Tests/PlotTests/Assertions.swift]: Perhaps additionally something like this to HTMLTests.swift

```
    func testPageClass() {
        let html = HTML(.class("no-js"))
        XCTAssertEqual(html.render(), #"<!DOCTYPE html><html class="no-js"></html>"#)
    }

    func testPageLanguageClass() {
        let html = HTML(.class("no-js"),.lang(.english))
        XCTAssertEqual(html.render(), #"<!DOCTYPE html><html class="no-js" lang="en"></html>"#)
    }
```

File: Sources/Plot/API/HTMLAttributes.swift [hunk 6]
```diff
@@ -133,11 +133,7 @@ public extension Attribute where Context == HTML.LinkContext {
     /// - parameter language: The language to assign.
     /// - parameter country: The country of the language to assign.
     static func hreflang(_ language: Language, _ country: Country? = nil) -> Attribute {
-        if let country = country {
-            return Attribute(name: "hreflang", value: "\(language.rawValue)-\(country.rawValue)")
-        } else {
-            return Attribute(name: "hreflang", value: language.rawValue)
-        }
+        Attribute(name: "hreflang", value: country != nil ? "\(language.rawValue)-\(country?.rawValue ?? "")": language.rawValue)
```
Comment (rpinz) [inline Sources/Plot/API/HTMLAttributes.swift]: Personally I prefer this variant over the former it replaces.  But the former is easier to quickly read I think.

File: Sources/Plot/API/Country.swift [hunk 7]
```diff
@@ -0,0 +1,260 @@
+/**
+*  Plot
+*  Copyright (c) John Sundell 2019
+*  MIT license, see LICENSE file for details
+*/
+
+// Enum defining country codes according to the ISO 3166-1 standard,
+// which can be used to specify the localized language of a page with
+// <link rel="alternate" href="http://fqdn.site/" hreflang="en-GB"> 
+public enum Country: String {
```
Comment (JohnSundell) [inline Sources/Plot/API/Country.swift]: The `Language` enum already includes some variants, like `.ukEnglish` and `.usEnglish`. Is this really needed, given that it lets you form invalid combinations, such as `en-SV`? I think extending `Language` with variants as needed is a better approach.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 6 above)
Comment (JohnSundell) [reply Sources/Plot/API/HTMLAttributes.swift]: Yeah, with so much optional unwrapping and string interpolation going on within a single long line, the code becomes really messy IMO.

File: Sources/Plot/API/HTMLAttributes.swift [hunk 8]
```diff
@@ -128,6 +128,31 @@ public extension Node where Context: HTMLLinkableContext {
...
+
+public extension Attribute where Context == HTML.LinkContext {
+    /// Assign an icon sizes string to this element.
+    /// - parameter sizes: The icon sizes of the link to assign.
+    static func sizes(_ sizes: String) -> Attribute {
+        Attribute(name: "sizes", value: sizes)
+    }
+}
+
+public extension Attribute where Context == HTML.LinkContext {
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: This extension can be merged with the above two ones. No need to create a new extension for each new API.

File: Tests/PlotTests/HTMLTests.swift [hunk 9]
```diff
@@ -128,6 +128,55 @@ final class HTMLTests: XCTestCase {
...
 
+    func testLinkHrefLangEnglish() {
+        let html = HTML(.head(.link(.rel(.alternate), .href("http://site/"), .hreflang(.english))))
+        assertEqualHTMLContent(html, """
+        <head><link rel="alternate" href="http://site/" hreflang="en"/></head>
+        """)
+    }
+
+    func testLinkHrefLangEnglishUS() {
+        let html = HTML(.head(.link(.rel(.alternate), .href("http://site/"), .hreflang(.english,.us))))
```
Comment (JohnSundell) [inline Tests/PlotTests/HTMLTests.swift]: Here for example you can use `.hreflang(.usEnglish)` to make this test pass, without introducing the `Country` enum.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 2 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: ok, I added "auEnglish" and "caEnglish" (Australia/Canada) that probably covers the most common English countries, removed the "Country.swift" file, and updated the test.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 6 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: This code has been removed, but I do have a question for you about it.  In this case country != nil tests that country is initialized and contains a value, what is your preference with "!"?  It would seem 100% safe to force unwrap over chaining, so that no default value is necessary.  I must admit I am in the habit of of never force unwrapping mostly because swiftlint complains, and it seems like it may create an edge case.  What rules to this do you apply when coding?

File: Sources/Plot/API/Country.swift (hunk 7 above)
Comment (rpinz) [reply Sources/Plot/API/Country.swift]: ok, I must admit here that I am probably not the best person to architect the localization as I have very little experience with languages other than English, specifically the US variation.  And I completely overlooked those 2 entries in Language.swift, so I removed the attempt at ISO 3166, and added caEnglish "en-ca", and auEnglish "en-au".

File: Tests/PlotTests/HTMLTests.swift (hunk 9 above)
Comment (rpinz) [reply Tests/PlotTests/HTMLTests.swift]: done.

File: Sources/Plot/API/HTMLAttributes.swift (hunk 8 above)
Comment (rpinz) [reply Sources/Plot/API/HTMLAttributes.swift]: done.

File: Sources/Plot/API/HTMLAttributes.swift [hunk 10]
```diff
@@ -128,6 +128,27 @@ public extension Node where Context: HTMLLinkableContext {
     }
 }
 
+public extension Attribute where Context == HTML.LinkContext {
+    /// Assign a hreflang string to this element.
+    /// - parameter language: The language to assign.
+    /// - parameter country: The country of the language to assign.
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: `country` parameter has been removed, so this is no longer needed.

File: Sources/Plot/API/HTMLAttributes.swift [hunk 11]
```diff
@@ -128,6 +128,27 @@ public extension Node where Context: HTMLLinkableContext {
     }
 }
 
+public extension Attribute where Context == HTML.LinkContext {
+    /// Assign a hreflang string to this element.
```
Comment (JohnSundell) [inline Sources/Plot/API/HTMLAttributes.swift]: This API is no longer string-based.

File: Sources/Plot/API/Language.swift [hunk 12]
```diff
@@ -190,6 +192,7 @@ public enum Language: String {
     case welsh = "cy"
     case wolof = "wo"
     case westernFrisian = "fy"
+    case xdefault = "x-default"
```
Comment (JohnSundell) [inline Sources/Plot/API/Language.swift]: I believe this should be camelCased as `xDefault`.

File: Tests/PlotTests/HTMLTests.swift [hunk 13]
```diff
@@ -128,6 +128,55 @@ final class HTMLTests: XCTestCase {
...
     }
 
+    func testLinkHrefLangEnglish() {
+        let html = HTML(.head(.link(.rel(.alternate), .href("http://site/"), .hreflang(.english))))
+        assertEqualHTMLContent(html, """
+        <head><link rel="alternate" href="http://site/" hreflang="en"/></head>
+        """)
+    }
+
+    func testLinkHrefLangEnglishUS() {
```
Comment (JohnSundell) [inline Tests/PlotTests/HTMLTests.swift]: I don't think this test is needed anymore, since it tests the exact same code path.

File: Tests/PlotTests/HTMLTests.swift [hunk 14]
```diff
@@ -128,6 +128,55 @@ final class HTMLTests: XCTestCase {
...
+        let html = HTML(.head(.link(.rel(.alternate), .href("http://site/"), .hreflang(.usEnglish))))
+        assertEqualHTMLContent(html, """
+        <head><link rel="alternate" href="http://site/" hreflang="en-us"/></head>
+        """)
+    }
+
+    func testLinkHrefLangXdefault() {
```
Comment (JohnSundell) [inline Tests/PlotTests/HTMLTests.swift]: Same here, we're not gaining any additional coverage by adding this test.

File: Tests/PlotTests/HTMLTests.swift [hunk 15]
```diff
@@ -128,6 +128,55 @@ final class HTMLTests: XCTestCase {
...
+        let html = HTML(.head(.link(.rel(.alternate), .href("http://site/"), .hreflang(.xdefault))))
+        assertEqualHTMLContent(html, """
+        <head><link rel="alternate" href="http://site/" hreflang="x-default"/></head>
+        """)
+    }
+
+    func testLinkSizes() {
```
Comment (JohnSundell) [inline Tests/PlotTests/HTMLTests.swift]: I think the follow test already tests this, so there's no need for this one.
